core/qthelper.cpp

```diff
@@ -1745,3 +1747,396 @@ extern "C" void unlock_planner()
+			res = '+' + res;
+		return fmt_string(res, flags, field_width, -1);
+	}
+
+	// If we have to prepend a '+' or a space character, remove that from the field width
+	char sign = 0;
+	if (i >= 0 && (flags.space || flags.sign) && field_width > 0) {
+		sign = flags.sign ? '+' : ' ';
+		--field_width;
+	}
+	if (flags.left)
+		field_width = -field_width;
+	QChar fillChar = flags.zero && !flags.left ? '0' : ' ';
+	QString res = QStringLiteral("%L1").arg(i, field_width, base, fillChar);
+	return sign ? insert_sign(res, sign) : res;
+}
+
+static QString fmt_float(double d, char type, vasprintf_flags flags, int field_width, int precision)
+{
+	// If we have to prepend a '+' or a space character, remove that from the field width
+	char sign = 0;
+	if (d >= 0.0 && (flags.space || flags.sign) && field_width > 0) {
+		sign = flags.sign ? '+' : ' ';
+		--field_width;
+	}
+	if (flags.left)
+		field_width = -field_width;
+	QChar fillChar = flags.zero && !flags.left ? '0' : ' ';
+	QString res = QStringLiteral("%L1").arg(d, field_width, type, precision, fillChar);
+	return sign ? insert_sign(res, sign) : res;
+}
+
+// Helper to extract integers from C-style format strings.
+// The default returned value, if no digits are found is 0.
+static int parse_fmt_int(const char **act)
+{
+	if (!isdigit(**act))
+		return 0;
+	int res = 0;
+	while (isdigit(**act)) {
+		res = res * 10 + **act - '0';
+		++(*act);
+	}
+	return res;
+}
+
+QString vasprintf_loc(const char *fmt, va_list ap)
+{
+	const char *act = fmt;
+	QString ret;
+	for (;;) {
+		// Get all bytes up to next '%' character and add them as UTF-8
+		const char *begin = act;
+		while (*act && *act != '%')
+			++act;
+		int len = act - begin;
+		if (len > 0)
+			ret += QString::fromUtf8(begin, len);
+
+		// We found either a '%' or the end of the format string
+		if (!*act)
+			break;
+		++act;	// Jump over '%'
+
+		if (*act == '%') {
+			++act;
+			ret += '%';
+			continue;
+		}
+		// Flags
+		vasprintf_flags flags = { 0 };
+		for (;; ++act) {
+			switch(*act) {
+			case '#':
+				flags.alternate_form = true;
+				break;
+			case '0':
+				flags.zero = true;
+				break;
+			case '-':
+				flags.left = true;
+				break;
+			case ' ':
+				flags.space = true;
+				break;
+			case '+':
+				flags.sign = true;
+				break;
+			case '\'':
+				flags.thousands = true;
+				break;
+			default:
+				goto flags_end;
+			}
+		}
+flags_end:
+		// Field width
+		int field_width;
+		if (*act == '*') {
+			field_width = va_arg(ap, int);
+			++act;
+		} else {
+			field_width = parse_fmt_int(&act);
+		}
+
+		// Precision
+		int precision = -1;
+		if (*act == '.') {
+			++act;
+			if (*act == '*') {
+				precision = va_arg(ap, int);
+				++act;
+			} else {
+				precision = parse_fmt_int(&act);
+			}
+		}
+
+		// Length modifier
+		enum {
+			LM_NONE,
+			LM_CHAR,
+			LM_SHORT,
+			LM_LONG,
+			LM_LONGLONG,
+			LM_LONGDOUBLE,
+			LM_INTMAX,
+			LM_SIZET,
+			LM_PTRDIFF
+		} length_modifier = LM_NONE;
```

Comment:
Not sure I love the enum declaration in the middle of the function... can we do that outside?